nucypher/blockchain/eth/signers/hardware.py

```diff
@@ -0,0 +1,243 @@
+        self._device_id = self.__client.get_device_id()
+        self.__addresses = dict()  # track derived addresses
+        self.__load_addresses()
+
+    #
+    # Internal
+    #
+
+    def __get_address_path(self, index: int = None, checksum_address: str = None) -> List[H_]:
+        """Resolves a checksum address into an HD path and returns it."""
+        if index is not None and checksum_address:
+            raise ValueError("Expected index or checksum address; Got both.")
+        elif index is not None:
+            hd_path = parse_path(f"{self.DERIVATION_ROOT}/{index}")
+        else:
+            try:
+                hd_path = self.__addresses[checksum_address]
+            except KeyError:
+                raise RuntimeError(f"{checksum_address} was not loaded into the device address cache.")
+        return hd_path
+
+    @handle_trezor_call
+    def __get_address(self, index: int = None, hd_path: Address = None, show_display: bool = True) -> str:
+        """Resolves a trezorlib HD path into a checksum address and returns it."""
+        if not hd_path:
+            if index is None:
+                raise ValueError("No index or HD path supplied.")  # TODO: better error handling here
+            hd_path = self.__get_address_path(index=index)
+        address = ethereum.__get_address(client=self.__client, n=hd_path, show_display=show_display)
+        return address
+
+    def __load_addresses(self) -> None:
+        """
+        Derive trezor addresses up to ADDRESS_CACHE_SIZE relative to
+        the calculated base path and internally cache them.
+        """
+        for index in range(self.ADDRESS_CACHE_SIZE):
+            hd_path = self.__get_address_path(index=index)
+            address = self.__get_address(hd_path=hd_path, show_display=False)
+            self.__addresses[address] = hd_path
+
+    @staticmethod
+    def _format_transaction(transaction_dict: dict) -> dict:
+        """
+        Handle Web3.py -> Trezor native transaction field formatting
+        # https://web3py.readthedocs.io/en/latest/web3.eth.html#web3.eth.Eth.sendRawTransaction
+        """
+        assert_valid_fields(transaction_dict)
+        trezor_transaction_keys = {'gas': 'gas_limit', 'gasPrice': 'gas_price', 'chainId': 'chain_id'}
+        trezor_transaction = dict(apply_key_map(trezor_transaction_keys, transaction_dict))
+        return trezor_transaction
+
+    @handle_trezor_call
+    def __sign_transaction(self, n: List[int], trezor_transaction: dict) -> Tuple[bytes, bytes, bytes]:
+        """Internal wrapper for trezorlib transaction signing calls"""
+        v, r, s = trezorlib.ethereum.sign_tx(client=self.__client, n=n, **trezor_transaction)
+        return v, r, s
+
+    #
+    # Trezor Signer API
+    #
+
+    @classmethod
+    def from_signer_uri(cls, uri: str) -> 'TrezorSigner':
+        """Return a trezor signer from URI string i.e. trezor:///my/trezor/path """
+        if uri != cls.URI_SCHEME:
```

Comment:
I guess you mean something like this?
```
        uri_breakdown = urlparse(uri)
        if uri_breakdown.scheme != cls.URI_SCHEME:
```

Reply:
It's a new situation - I am trying to support both URIs that include a complex scheme, and also simply just a keyword, for example:

`--signer keystore://<PATH>` & `--signer trezor`.

What might a trezor URI be ( I suppose it can contain the device ID - bu that seems unnecessary)?

Reply:
But the URI can be a trezor with a derivation path, no?

Reply:
Great idea! :bulb:  that had not occurred to me.  How might this look?